source/mesh_refinement/nonadiabatic_temperature_threshold.cc

```diff
@@ -0,0 +1,157 @@
+    {
+      // tag_additional_cells is executed before the equations are solved
+      // for the very first time. If we do not have the finite element, we
+      // do not have the temperature and just do nothing in this plugin.
+      if (this->get_dof_handler().n_locally_owned_dofs() == 0)
+        return;
+
+      const Quadrature<dim> quadrature(this->get_fe().base_element(this->introspection().base_elements.temperature).get_unit_support_points());
+      FEValues<dim> fe_values (this->get_mapping(),
+                               this->get_fe(),
+                               quadrature,
+                               update_quadrature_points | update_values);
+
+      std::vector<double> temperature_values (quadrature.size());
+      const unsigned int n_dofs_per_cell = this->get_fe().base_element(this->introspection().base_elements.temperature).dofs_per_cell;
+
+      for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+        if (cell->is_locally_owned())
+          {
+            bool refine = false;
+
+            fe_values.reinit(cell);
+            fe_values[this->introspection().extractors.temperature].get_function_values (this->get_solution(),
+                                                                                         temperature_values);
+
+            // if the nonadiabatic temperature exceeds the threshold, cell is marked for refinement
+            for (unsigned int j=0; j<n_dofs_per_cell; ++j)
+              {
+                const double adiabatic_temperature = this->get_adiabatic_conditions().temperature(fe_values.quadrature_point(j));
+
+                double nonadiabatic_temperature = 0;
+                if (temperature_anomaly_type == absolute_value)
+                  nonadiabatic_temperature = std::abs(temperature_values[j] - adiabatic_temperature);
+                else if (temperature_anomaly_type == positive_only)
+                  nonadiabatic_temperature = temperature_values[j] - adiabatic_temperature;
+                else if (temperature_anomaly_type == negative_only)
+                  nonadiabatic_temperature = adiabatic_temperature - temperature_values[j];
+                else
+                  AssertThrow (false, ExcNotImplemented());
+
+                if (nonadiabatic_temperature > threshold)
+                  {
+                    refine = true;
+                    break;
+                  }
+              }
+
+            if (refine)
+              {
+                cell->clear_coarsen_flag ();
+                cell->set_refine_flag ();
+              }
+          }
+    }
+
+    template <int dim>
+    void
+    NonadiabaticTemperatureThreshold<dim>::
+    declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Mesh refinement");
+      {
+        prm.enter_subsection("Nonadiabatic temperature threshold");
+        {
+          prm.declare_entry ("Threshold",
+                             "0",
```

Comment:
The code looks ready to merge, but I think we want to think about this threshold. '0' in the current setting means everything is always refined as soon as this criterion is selected. Would we prefer a setting where nothing is ever refined if this plugin is selected without changing this parameter? I think that would be easier to understand for new users, but I am open for discussion.
That would then mean:
```suggestion
                             boost::lexical_cast<std::string>(std::numeric_limits<double>::max()),
```

Reply:
I would argue refinishing nothing or everything are not good defaults. We could also leave the default empty so the user is forced to specify it?

Reply:
I like that idea. I've changed the code to make the default empty.

Reply:
Update: Leaving it empty does not work (see failed tests below; there might be a way to make it work, but I think it would be more complicated). So I instead changed it to 100 K, which seems like a reasonable default to me.